File: lib/models/session.dart
class Session {
  final String id;
  String name;
  final DateTime timestamp;
  DateTime lastModified;
  bool isIncognito;

  Session({
    required this.id,
    required this.name,
    required this.timestamp,
    DateTime? lastModified,
    this.isIncognito = false,
  }) : lastModified = lastModified ?? timestamp;

  factory Session.fromJson(Map<String, dynamic> json) {
    return Session(
      id: json['id'],
      name: json['name'],
      timestamp: DateTime.parse(json['timestamp']),
      lastModified: json.containsKey('lastModified')
          ? DateTime.parse(json['lastModified'])
          : DateTime.parse(json['timestamp']),
      isIncognito: json['isIncognito'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'timestamp': timestamp.toIso8601String(),
      'lastModified': lastModified.toIso8601String(),
      'isIncognito': isIncognito,
    };
  }
}

File: models/session/Session.ts
export interface Session {
  id: string;
  name: string;
  timestamp: Date;
  lastModified: Date;
  isIncognito: boolean;
}

export interface SessionJSON {
  id: string;
  name: string;
  timestamp: string;
  lastModified?: string;
  isIncognito?: boolean;
}

export const sessionToJSON = (session: Session): SessionJSON => {
  return {
    id: session.id,
    name: session.name,
    timestamp: session.timestamp.toISOString(),
    lastModified: session.lastModified.toISOString(),
    isIncognito: session.isIncognito,
  };
};

export const sessionFromJSON = (json: SessionJSON): Session => {
  return {
    id: json.id,
    name: json.name,
    timestamp: new Date(json.timestamp),
    lastModified: json.lastModified
      ? new Date(json.lastModified)
      : new Date(json.timestamp),
    isIncognito: json.isIncognito ?? false,
  };
};

export const createSession = (params: {
  id: string;
  name: string;
  timestamp: Date;
  lastModified?: Date;
  isIncognito?: boolean;
}): Session => {
  return {
    id: params.id,
    name: params.name,
    timestamp: params.timestamp,
    lastModified: params.lastModified ?? params.timestamp,
    isIncognito: params.isIncognito ?? false,
  };
};
